src/OperationProcessor.ts

```diff
@@ -427,11 +467,134 @@ class DidCacheImpl implements DidCache {
+    // operations which do have parents.
+    if (opInfo.type === OperationType.Create) {
+      this.processCreateOperation(opHash, opInfo);
+    } else {
+      this.processOperationWithParent(opHash, opInfo);
+    }
+
+    // Process operations waiting on this operation
+    this.processDescendantsWaitingOn(opHash);
+  }
+
+  private processCreateOperation (_opHash: OperationHash, opInfo: OperationInfo): void {
+    // TODO: Validate create operation (verify signature)
+    opInfo.status = OperationStatus.Valid;
+  }
+
+  private processOperationWithParent (opHash: OperationHash, opInfo: OperationInfo): void {
+    const parentOpHash = opInfo.parent as OperationHash;
+    const parentOpInfo = this.opHashToInfo.get(parentOpHash);
+
+    // If we do not know about the parent, then the ancestry of this operation is
+    // incomplete. We leave the status to be Unvalidated and update the waitingDescendants
+    // of the parent operation (hash).
+    if (parentOpInfo === undefined) {
+      // assert: opInfo.statue === OperationStatus.Unvalidated
+      opInfo.missingAncestor = parentOpHash;
+      this.updateWaitingDescendants(opInfo.missingAncestor, opHash);
+      return;
+    }
+
+    // The parent has an Unvalidated status. This implies an incomplete ancestry of the
+    // parent and therefore of this operation. We leave the status to be Unvalidated and
+    // update the waitingDescendants of the closest missing ancestor.
+    if (parentOpInfo.status === OperationStatus.Unvalidated) {
+      opInfo.missingAncestor = parentOpInfo.missingAncestor as OperationHash;
+      this.updateWaitingDescendants(opInfo.missingAncestor, opHash);
+      return;
+    }
+
+    // If the parent is invalid, then this operation is invalid as well.
+    if (parentOpInfo.status === OperationStatus.Invalid) {
+      opInfo.status = OperationStatus.Invalid;
+      return;
+    }
+
+    // Assert: parentOpInfo.status === OperationStatus.Valid. Validate the operation
+    if (!this.validate(opHash, opInfo)) {
+      opInfo.status = OperationStatus.Invalid;
+      return;
+    }
+
+    // The operation is intrinsically valid. Before we set it to valid, we need
+    // to ensure that it is the earliest sibling among child nodes of its parent.
+    const curEarliestSiblingHash = this.nextVersion.get(parentOpHash);
+
+    if (curEarliestSiblingHash !== undefined) {
+      const curEarliestSiblingInfo = this.opHashToInfo.get(curEarliestSiblingHash) as OperationInfo;
+      if (earlier(curEarliestSiblingInfo.timestamp, opInfo.timestamp)) {
+        opInfo.status = OperationStatus.Invalid;
+        return;
+      } else {
+        this.invalidatePreviouslyValidOperation(curEarliestSiblingHash);
+        // fall through ...
+      }
+    }
+
+    opInfo.status = OperationStatus.Valid;
+    this.nextVersion.set(parentOpHash, opHash);
+    return;
+  }
+
+  private updateWaitingDescendants (missingAncestor: OperationHash, opHash: OperationHash) {
```

Comment:
Consider renaming method to "addWaitingDescendant(...), seems more accurate.

Reply:
Good suggestion - incorporated it.